### PR TITLE
Add a controller which creates a tekton pipeline run from a prow job

### PR DESCRIPTION
**Pull request by ccojocar**
This controller is able to start a Tekton pipeline using the embedded PipelineRun specification defined in the prow job. The user will have to install beforehand all the resources referenced by this Tekton PipelineRun (such as Pipeline CRD, Task CRDs and PipelineResource CRDs).

Before starting the pipeline, the controller will create dynamically the git PipelineResoruce from prow job git reference and it will add it to the PipelineRun input parameters. 

[Design Doc](https://docs.google.com/document/d/1Lsp4QjzZLev4T2RttHV9srg0y1RiSEiyBjL9sFR7aKg/edit?usp=sharing)

cc @rawlingsj @wbrefvem

### CODE REVIEW

**Comment by ccojocar** [comment]
cc @fejta

**Comment by stevekuznetsov** [comment]
Are there design documents for this? I would love to have a little more rigor for adding 11k LoC.

**Comment by spiffxp** [comment]
/hold
I agree, I'd really like to see some docs around this.  I appreciate the contribution, but IMO dropping a massive PR with no initial conversation or documentation is not the path to a smooth landing in open source projects

File: prow/cmd/pipeline/controller.go [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+		case apierrors.IsNotFound(err):
+			// Do not have a pipeline
+		case err != nil:
+			return fmt.Errorf("get pipelinerun %s: %v", key, err)
+		}
+		if p != nil {
+			havePipelineRun = true
+		}
+	case pipelineRun:
+		p, err = c.getPipelineRun(ctx, namespace, name)
```
Comment (fejta) [inline prow/cmd/pipeline/controller.go]: Ideally this should support creating pipelines in foreign clusters, like plank and the build controllers do

**Comment by ccojocar** [comment]
My understanding was that @rawlingsj had some discussions with @cjwagner and @fejta before initialising this work.  Anyhow, I will try to prepare a design document hopefully by the next sig-testing meeting that everyone is on the same page.

**Comment by ccojocar** [comment]
This is the [design proposal ](https://docs.google.com/document/d/1Lsp4QjzZLev4T2RttHV9srg0y1RiSEiyBjL9sFR7aKg/edit?usp=sharing)which I would like to add it on today's sig-testing meeting agenda.

File: prow/cmd/pipeline/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,112 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")
+load("@io_bazel_rules_k8s//k8s:object.bzl", "k8s_object")
+load("//prow:def.bzl", "prow_image")
+
+k8s_object(
+    name = "dev",
+    cluster = "gke_fejta-prod_us-central1-f_erick",
```
[the remark this answers is not in the document]
Comment (ccojocar) [reply prow/cmd/pipeline/BUILD.bazel]: Yeah, I can remove it.

File: prow/cmd/pipeline/controller.go (hunk 1 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I can make this multi-cluster aware.

**Comment by ccojocar** [comment]
In the design proposal review was decided to split this functionality in two separate controllers. The controller from this PR which starts directly the Tekton pipeline and another dedicated controller for Jenkins X, that will convert the Jenkins X pipeline file into Tektont CRDs and then start the appropriate pipeline. The Jenkins X controller, it will be implemented in a separate PR.

**Comment by ccojocar** [comment]
I had to update the `knative/pkg` and `knative/build` to latest version in order to be able to use the latest `tektoncd/pipeline` version [v0.2.0](https://github.com/tektoncd/pipeline/releases). I sent these changes also in a separate PR https://github.com/kubernetes/test-infra/pull/12008.

File: prow/cmd/pipeline/controller.go (hunk 1 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: The current implementation supports creating pipelines in different clusters. It keeps a map between the cluster context and the pipelineConfig, in a similar way like the build controller. Every time a  ProwJob/PipelineRun is created, it looks up the cluster config based on the given context.

**Comment by fejta** [comment]
Can we rebase this commit now that the other PR merged?

**Comment by ccojocar** [comment]
@fejta it is now rebased. It would be great, if you could review it. I am trying to validate it with [tektoncd/pipeline](https://github.com/tektoncd/pipeline/releases/tag/0.2.0) version `v0.2.0`.

File: go.mod [hunk 3]
```diff
@@ -75,6 +77,7 @@ require (
 	github.com/sirupsen/logrus v1.2.0
 	github.com/spf13/cobra v0.0.3
 	github.com/spf13/pflag v1.0.3
+	github.com/tektoncd/pipeline v0.1.0
```
Comment (fejta) [inline go.mod]: Maybe move this out of this commit?

File: prow/apis/prowjobs/v1/types.go [hunk 4]
```diff
@@ -71,6 +72,8 @@ const (
 	JenkinsAgent ProwJobAgent = "jenkins"
 	// KnativeBuildAgent means prow will schedule the job via a build-crd resource.
 	KnativeBuildAgent ProwJobAgent = "knative-build"
+	// TektonAgent means prow will schedule the job via a tekton PipelineRun CRD resource.
+	TektonAgent = "tekton-pipeline"
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types.go]: Would just `tekton` be sufficient? Or do we need the `pipeline` suffix as well?

File: prow/cmd/pipeline/admission.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	response, err := decide(ar)
+	if err != nil {
+		logrus.WithError(err).Error("failed decision")
+		response = &admissionapi.AdmissionResponse{
+			Result: &meta.Status{
+				Message: err.Error(),
+			},
+		}
+	}
+	var result admissionapi.AdmissionReview
```
Comment (fejta) [inline prow/cmd/pipeline/admission.go]: Can we drop the admission controller part? I think that's a rather weak part of the knative-build controller

Instead I think we should write a prowjob admission controller that validates a prowjob spec does not change. We don't need one for each type of agent.

File: prow/cmd/pipeline/controller.go (hunk 1 above)
Comment (fejta) [reply prow/cmd/pipeline/controller.go]: Oh I see! Can you choose a different name for the kubectl context than `ctx`? I was thinking this was the `context.Context` builtin.

File: prow/cmd/pipeline/controller.go [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+
+// pipelineMeta builds the pipeline metadata from prow job definition
+func pipelineMeta(pj prowjobv1.ProwJob) metav1.ObjectMeta {
+	labels, annotations := decorate.LabelsAndAnnotationsForJob(pj)
+	return metav1.ObjectMeta{
+		Annotations:     annotations,
+		Name:            pj.Name,
+		Namespace:       pj.Spec.Namespace,
+		Labels:          labels,
+		OwnerReferences: []metav1.OwnerReference{ownerRef(pj)},
```
Comment (fejta) [inline prow/cmd/pipeline/controller.go]: Can we please drop the owner references? This does not work with cross cluster resources. AKA if you create a resource in a foreign cluster with an owner reference to a local resource, the foreign cluster will immediately delete the object because resource name does not exist in the foreign cluster.

I'm pretty sure later commits of the build controller dropped owner references, this is why

File: prow/apis/prowjobs/v1/types.go (hunk 4 above)
Comment (ccojocar) [reply prow/apis/prowjobs/v1/types.go]: I am fine either way.

File: prow/cmd/pipeline/admission.go (hunk 5 above)
Comment (ccojocar) [reply prow/cmd/pipeline/admission.go]: I removed entirely the admission controller and also the cert/key arguments from the controller command line.

Do you agree to implement the generic prowjob admission controller in a separate PR?

File: prow/cmd/pipeline/controller.go (hunk 6 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I removed it. This was my addition, I guess. I didn't know about this behaviour cross clusters. Thanks for explaining the details.

**Comment by ccojocar** [comment]
@fejta I would like to add some docs and a deployment file in a follow up PR. Where is the best place for the documentation? is it fine to create a `tekton.md` file into the `prow` folder?

**Comment by stevekuznetsov** [comment]
Did we finalize the design doc and can we identify which part of that doc this PR is fulfilling?

File: prow/cmd/pipeline/dev.yaml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  template:
+    metadata:
+      labels:
+        app: prow-pipeline
+    spec:
+      serviceAccount: prow-pipeline
+      containers:
+      - name: pipeline
+        image: gcr.io/k8s-testimages/pipeline:latest  # Note: not gcr.io/k8s-prow for dev
+        imagePullPolicy: Always  # Good practice for dev/debugging, bad for prod
```
Comment (stevekuznetsov) [inline prow/cmd/pipeline/dev.yaml]: This is great for prod when using a floating tag like `latest` if you want updates to percolate and ensure node image caches are not making prod deployments change based on target node

File: prow/cmd/pipeline/main.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+	var o options
+	if err := o.parse(flag.CommandLine, os.Args[1:]); err != nil {
+		logrus.Fatalf("Invalid flags: %v", err)
+	}
+	return o
+}
+
+func (o *options) parse(flags *flag.FlagSet, args []string) error {
+	flags.BoolVar(&o.allContexts, "all-contexts", false, "Monitor all cluster contexts, not just default")
+	flags.StringVar(&o.totURL, "tot-url", "", "Tot URL")
```
Comment (stevekuznetsov) [inline prow/cmd/pipeline/main.go]: Why do we want to support this? Isn't `tot` deprecated?

File: prow/pod-utils/downwardapi/jobspec.go [hunk 9]
```diff
@@ -177,3 +177,15 @@ func GetRevisionFromSpec(spec *JobSpec) string {
...
 }
+
+// GetBranch returns the source code branch
+func GetBranch(spec *JobSpec) string {
+	branch := spec.Refs.BaseRef
+	if spec.Type == prowapi.PostsubmitJob || spec.Type == prowapi.BatchJob {
+		return branch
+	}
+	if len(spec.Refs.Pulls) > 0 {
+		branch = fmt.Sprintf("PR-%v", spec.Refs.Pulls[0].Number)
```
Comment (stevekuznetsov) [inline prow/pod-utils/downwardapi/jobspec.go]: What is this?

**Comment by fejta** [comment]
> Where is the best place for the documentation? is it fine to create a tekton.md file into the prow folder?

How about prow/cmd/pipeline/README.md with a link to it somewhere on the main prow readme and/or jobs.md?

> Did we finalize the design doc and can we identify which part of that doc this PR is fulfilling?

Seems like we're all in agreement to separate out the raw pipeline run agent from the jenkins-x agent. This implements the pipeline-run agent and removes the jenkins-x agent.

File: prow/cmd/pipeline/dev.yaml [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        - containerPort: 8443
+          name: validator-http
+          protocol: TCP
+        volumeMounts:
+        - name: tls
+          mountPath: /etc/tls
+      volumes:
+      - name: tls
+        secret:
+          secretName: prow-pipeline
```
Comment (cjwagner) [inline prow/cmd/pipeline/dev.yaml]: I think we can get rid of the ports, volume, and tls args now that the admission controller is omitted.

File: prow/cmd/pipeline/dev.yaml [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  - port: 443
+    protocol: TCP
+    targetPort: 8443
+  selector:
+    app: prow-pipeline
+  type: ClusterIP
+---
+
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
```
Comment (cjwagner) [inline prow/cmd/pipeline/dev.yaml]: We can get rid of this and the service above now too.

File: prow/cmd/pipeline/main.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+	flags.StringVar(&o.kubeconfig, "kubeconfig", "", "Path to kubeconfig. Only required if out of cluster")
+	flags.StringVar(&o.config, "config", "", "Path to prow config.yaml")
+	flags.StringVar(&o.pipelineCluster, "pipeline-cluster", "", "Path to file containing a YAML-marshalled kube.Cluster object. If empty, uses the local cluster.")
```
Comment (cjwagner) [inline prow/cmd/pipeline/main.go]: This should still be called "build-cluster". The naming is bad, but this doesn't refer to knative builds, rather the cluster(s) where we execute workloads. We use the `--build-cluster` flag across Prow, not just for the Prow's knative build controller. e.g. https://github.com/kubernetes/test-infra/blob/0ffe02f69264e74009ef3856dd746f54961af2b1/prow/cmd/plank/main.go#L65

File: prow/cmd/pipeline/main.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+		}
+		pipelineConfigs[context] = *bc
+	}
+
+	controller, err := newController(kc, pjc, pjif.Prow().V1().ProwJobs(), pipelineConfigs, o.totURL, configAgent.Config, kube.RateLimiter(controllerName))
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating controller")
+	}
+
+	if err := controller.Run(1, stop); err != nil {
```
Comment (cjwagner) [inline prow/cmd/pipeline/main.go]: Why only 1 worker thread? The build controller uses 2 (also why?).

File: prow/cmd/pipeline/controller.go [hunk 14]
```diff
@@ -0,0 +1,673 @@
...
+		return nil, err
+	}
+	return pc.client.TektonV1alpha1().PipelineResources(namespace).Create(pr)
+}
+
+func (c *controller) now() metav1.Time {
+	return metav1.Now()
+}
+
+func (c *controller) pipelineID(pj prowjobv1.ProwJob) (string, error) {
```
Comment (cjwagner) [inline prow/cmd/pipeline/controller.go]: This function doesn't seem right to me. In particular it will always fail for periodic jobs because periodics always have nil `Spec.Refs`.
Why was this changed from what the build controller does? https://github.com/kubernetes/test-infra/blob/b7b7a8907608cc78abae88860367eb6710157aa5/prow/cmd/build/controller.go#L345-L352

File: prow/cmd/pipeline/controller.go [hunk 15]
```diff
@@ -0,0 +1,673 @@
...
+				runtime.HandleError(fmt.Errorf("failed to reconcile %s: %v", key, err))
+				return // Do not forget so we retry later.
+			}
+			c.workqueue.Forget(key)
+		}()
+	}
+}
+
+// toKey returns context/namespace/name/kind
+func toKey(ctx, namespace, name, kind string) string {
```
Comment (cjwagner) [inline prow/cmd/pipeline/controller.go]: Instead of including the kind in the key so that we can distinguish ProwJobs from Pipelines can we instead always name the Pipeline with the same name as the ProwJob and then look up both the ProwJob and the Pipeline when reconciling? This is what we do in the build controller: https://github.com/kubernetes/test-infra/blob/b7b7a8907608cc78abae88860367eb6710157aa5/prow/cmd/build/controller.go#L397-L414
Plank also creates pods with names matching the owning ProwJob. This is very helpful for debugging and is assumed in a number of places so we should try to avoid breaking the assumption.

File: prow/cmd/pipeline/admission.go (hunk 5 above)
Comment (ccojocar) [reply prow/cmd/pipeline/admission.go]: I think this is the goal of this issue https://github.com/kubernetes/test-infra/issues/8468.

File: prow/cmd/pipeline/main.go (hunk 8 above)
Comment (ccojocar) [reply prow/cmd/pipeline/main.go]: What is the new alternative to generate build numbers?

File: prow/pod-utils/downwardapi/jobspec.go (hunk 9 above)
Comment (ccojocar) [reply prow/pod-utils/downwardapi/jobspec.go]: This was a branch made up for PR, in order to allow `tot` to generate a build number per PR. 

I will refactor it as suggested by @cjwagner bellow.

File: prow/cmd/pipeline/main.go (hunk 13 above)
Comment (ccojocar) [reply prow/cmd/pipeline/main.go]: In my understanding, having `2` worker threads ensures that two reconcile loops can run in parallel. This means that 2 items from the queue will be processed in parallel.  It will likely lead to lower latency since the build controller handles 2 types of CRDs (ProwJob and Build). In the same time, can lead to object update conflicts when the subsequent items are of the same type. For instance, one worker has already updated the ProwJob object while the other one is processing the previous version from the cache. Maybe a PATCH update strategy can mitigate this. 

I can also set 2 worker threads for pipeline controller.

File: prow/cmd/pipeline/controller.go (hunk 14 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I think, we wanted to generate build number per branch instead of prow job which for row pipeline scenario is not relevant. 

I will use the same logic like in the build controller to generate the build IDs.

File: prow/cmd/pipeline/controller.go (hunk 15 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: The PipelineRun is created with the same name as the ProwJob, but it has has a label with the ProwJob ID, which we used as a selector.

The reason for these is that in Jenkins X, we are using a different name for the pipeline run that encodes the repository + branch names, that is not the case anymore for raw pipeline use case.

If you consider that this change is really necessary, I can refactor the code.

File: prow/cmd/pipeline/dev.yaml (hunk 7 above)
Comment (ccojocar) [reply prow/cmd/pipeline/dev.yaml]: This deployment is intended only to be used during development. That's why is using `Always` with `latest`. 

I will follow up with a production deployment in a different PR with will use a tag.

File: prow/pod-utils/downwardapi/jobspec.go (hunk 9 above)
Comment (ccojocar) [reply prow/pod-utils/downwardapi/jobspec.go]: removed

File: prow/cmd/pipeline/dev.yaml (hunk 10 above)
Comment (ccojocar) [reply prow/cmd/pipeline/dev.yaml]: done

File: prow/cmd/pipeline/dev.yaml (hunk 11 above)
Comment (ccojocar) [reply prow/cmd/pipeline/dev.yaml]: done

File: prow/cmd/pipeline/main.go (hunk 12 above)
Comment (ccojocar) [reply prow/cmd/pipeline/main.go]: done

File: prow/cmd/pipeline/main.go (hunk 13 above)
Comment (ccojocar) [reply prow/cmd/pipeline/main.go]: Set it to 2 worker threads.

File: prow/cmd/pipeline/controller.go (hunk 14 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: used the same way to generate the build ID like in the build controller

File: prow/cmd/pipeline/controller.go (hunk 15 above)
Comment (stevekuznetsov) [reply prow/cmd/pipeline/controller.go]: Yes please lets keep this identical between Prow controllers

File: prow/cmd/pipeline/main.go (hunk 13 above)
Comment (stevekuznetsov) [reply prow/cmd/pipeline/main.go]: I don't think multiple versions of the same object are in the queue for the controller at the same time, keys are coalesced IIRC.

File: prow/cmd/pipeline/controller.go (hunk 15 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I removed the `kind` from the key and refactored the code to follow the same logic like the build controller.

**Comment by stevekuznetsov** [comment]
I don't have time this week to review and @cjwagner is on vacation, so @fejta if you're comfortable with your review, feel free to merge once this is rebased

**Comment by ccojocar** [comment]
I am still doing some validation today. I'll let you know when is ready to be merged.

**Comment by fejta** [comment]
sgtm

**Comment by fejta** [comment]
Please `/hold cancel` when you want this to have another review

**Comment by ccojocar** [comment]
@fejta Please could you have a final review. I validated with latest tekton from master, and it works fine so far. 

I had to remove the `DefaultClusterAlias` from pipeline config map, because the `InClusterContext` and `DefaultClusterAlias` is removing each other pipelines since they point to the same cluster context. I think the solution is to map the DefaultClusterAlias -> InClusterContext when enqueuing the keys. 

The same issue occurs also in the build controller, I'll send the fix for it in a different PR.

File: prow/cmd/pipeline/controller.go [hunk 16]
```diff
@@ -192,6 +192,14 @@ func newController(opts controllerOptions) (*controller, error) {
 	return c, nil
 }
 
+// clusterToCtx converts the prow job's cluster to a cluster context
+func clusterToCtx(cluster string) string {
+	if cluster == kube.DefaultClusterAlias {
+		return kube.InClusterContext
```
Comment (fejta) [inline prow/cmd/pipeline/controller.go]: This is not always true

File: prow/apis/prowjobs/v1/types.go [hunk 17]
```diff
@@ -71,6 +72,8 @@ const (
 	JenkinsAgent ProwJobAgent = "jenkins"
 	// KnativeBuildAgent means prow will schedule the job via a build-crd resource.
 	KnativeBuildAgent ProwJobAgent = "knative-build"
+	// TektonAgent means prow will schedule the job via a tekton PipelineRun CRD resource.
+	TektonAgent = "tekton"
```
Comment (fejta) [inline prow/apis/prowjobs/v1/types.go]: tekton-pipeline? Seems weird for agent: tekton to require a pipeline_run_spec

File: prow/config/config.go [hunk 18]
```diff
@@ -1123,7 +1123,8 @@ func validateAgent(v JobBase, podNamespace string) error {
 	k := string(prowapi.KubernetesAgent)
 	b := string(prowapi.KnativeBuildAgent)
 	j := string(prowapi.JenkinsAgent)
-	agents := sets.NewString(k, b, j)
+	p := string(prowapi.TektonAgent)
```
Comment (fejta) [inline prow/config/config.go]: Can we validate that we can set Namespace when using pipelines?

File: prow/apis/prowjobs/v1/types.go (hunk 17 above)
Comment (ccojocar) [reply prow/apis/prowjobs/v1/types.go]: `tekton-pipeline` was the original name, but I can rename it back.

Comment (ccojocar) [reply prow/apis/prowjobs/v1/types.go]: I renamed the agent to `tekton-pipeline`.

File: prow/config/config.go (hunk 18 above)
Comment (ccojocar) [reply prow/config/config.go]: I fixed the config to allow namespace customisation for `tekton-pipeline` agent, and validated.

File: prow/cmd/pipeline/controller.go (hunk 16 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: Which is the use case when this is not true?

Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: It seems that there is an issue when both `DefaultClusterAlias` and `InClusterContext` point to the same cluster (this is the default). The PipelineRun resource created for a prow job with `spec.Cluster: default`, is removed by the `InClusterContext` reconciliation because `spec.Cluster: default != spec.Cluster: ""`

**Comment by ccojocar** [comment]
@fejta I addressed this issues. You can cancel the hold if the changes look fine to you. Thanks

File: prow/cmd/pipeline/controller.go (hunk 16 above)
Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I explained the issue here for build controller https://github.com/kubernetes/test-infra/pull/12177#discussion_r275317270

Comment (fejta) [reply prow/cmd/pipeline/controller.go]: This seems more appropriate for the other PR. I'd recommend taking out the commit from here as I don't think it works correctly. But happy to address it in a follow-up

Comment (ccojocar) [reply prow/cmd/pipeline/controller.go]: I removed that commit from this PR. We'll work out a solution in a follow up PR. Thanks
